### PR TITLE
Replace Dijkstra

### PR DESCRIPTION
**Pull request by Sne4kers**
- Replaced the "Dijkstra all" algorithm with an algorithm specific to CFG implementation
- Added a test set to test the new algorithm
- Added getRotatedAdjacencyList precalculation in the algorithm to improve its performance. I think it is a good thing to refactor CFG in terms of the way it is stored, so we can gain performance benefits in the future. Right now I have an idea of changing the Node interface, so it would have a list of children and a list of parents, so we can easier traverse the CFG in the future. But I need a bit more time to think if it's is a good solution for this project.

### CODE REVIEW

File: src/analysis/static/graph/CFG.ts [hunk 1]
```diff
@@ -50,4 +51,70 @@ export class CFG {
       .filter((node) => node.type === NodeType.Root)
       .map((node) => <RootNode>node);
   }
+
+  getNodeById(nodeId : string): Node {
+    const found = this._nodes.filter((node: Node) => node.id == nodeId);
+    if(found.length != 1) {
```
Comment (dstallenberg) [inline src/analysis/static/graph/CFG.ts:57]: If you use array.find you can simply use if(found) to check if found is non null.
I also suggest to rename found -> node.

File: src/analysis/static/graph/CFG.ts [hunk 2]
```diff
@@ -50,4 +51,70 @@ export class CFG {
...
+      console.log("No node with such id in CFG");
```
Comment (dstallenberg) [inline src/analysis/static/graph/CFG.ts:58]: Please refrain from using console.log.
This function should either throw an error when it is not found or simply return null.
in the latter case we do not even need an if-statement.

File: src/analysis/static/graph/CFG.ts [hunk 3]
```diff
@@ -50,4 +51,70 @@ export class CFG {
...
+      return null;
+    }
+    return found[0];
+  }
+
+  getRotatedAdjacencyList(): Map<String, Pair<string, number>[]> {
```
Comment (dstallenberg) [inline src/analysis/static/graph/CFG.ts:64]: Can you maybe at some documentation as to what the map exactly contains.
I'm guessing that the "second" in the pair is the weight?

File: src/analysis/static/graph/CFG.ts [hunk 4]
```diff
@@ -50,4 +51,70 @@ export class CFG {
...
+          "first" : edge.from, 
+          "second" : edge.branchType !== undefined ? 1 : 0
+        }
+      );
+    }
+
+    return adjList;
+  }
+
+  findClosestAncestor(from: string, targets: string[]): [number, Node] {
```
Comment (dstallenberg) [inline src/analysis/static/graph/CFG.ts:85]: I would suggest to change targets type to Set<string> immediately instead of converting inside the method.

Comment (dstallenberg) [inline src/analysis/static/graph/CFG.ts:85]: When returing multiple things in a method refrain from returning an array.
Instead return {distance: number, ancestor: Node}.
This way we can use destructuring when calling the function
Example:
const {ancestor} = findClosestAncestor(...)

File: src/analysis/static/graph/CFG.ts [hunk 5]
```diff
@@ -50,4 +51,70 @@ export class CFG {
...
+    let targetsSet = new Set<string>(targets);
+    const rotatedAdjList = this.getRotatedAdjacencyList();
+
+    let visitedNodeIdSet = new Set<string>([from]);
+    const searchQueue = [];
```
Comment (dstallenberg) [inline src/analysis/static/graph/CFG.ts:90]: Missing type

File: src/analysis/static/graph/CFG.ts [hunk 6]
```diff
@@ -50,4 +51,70 @@ export class CFG {
...
+        // return if of targets nodes was found
+        if(targetsSet.has(nextNodeId)) {
+          return [currentDistance + (pairOfParent.second), this.getNodeById(nextNodeId)];
+        }
+        // add element to queue and visited nodes to continue search
+        visitedNodeIdSet.add(nextNodeId);
+        searchQueue.push([currentDistance + (pairOfParent.second), nextNodeId]);
+      }
+    }
+    console.log("No covered nodes found");
```
Comment (dstallenberg) [inline src/analysis/static/graph/CFG.ts:117]: Remove console.log please

Comment (Sne4kers) [reply src/analysis/static/graph/CFG.ts]: Fixed

**Review by dstallenberg — APPROVED** [review]

**Pull request merged**